File: AtCoder/ABC/179/python/E_Sequence_Sum.py
N, X, M = map(int, input().split())

ordinal = [-1] * M
rec = []
cyc = []
res = 0
for i in range(N):
    if ordinal[X] != -1:
        cyc = rec[ordinal[X]:i]
        break
    ordinal[X] = i
    rec.append(X)
    res += X
    X = X * X % M
N -= len(rec)
if N == 0:
    print(res)
    exit()
freq = len(cyc)
q = N // freq
r = N % freq
_sum = [0] * (freq + 1)
for i in range(freq):
    _sum[i + 1] = _sum[i] + cyc[i]
res += _sum[freq] * q + _sum[r]
print(res)

File: AtCoder/ABC/179/cpp/E_Sequence_Sum.cpp
#include <iostream>
#include <stdio.h>
#include <vector>

using namespace std;
typedef long long ll;

ll N, X, M;

int main() {
  cin >> N >> X >> M;
  vector<ll> ord(M, -1);
  vector<ll> rec, cyc;
  ll res = 0;
  for (ll i = 0; i < N; ++i) {
    if (ord[X] != -1) {
      for (ll j = ord[X]; j < i; ++j) {
        cyc.push_back(rec[j]);
      }
      break;
    }
    ord[X] = i;
    rec.push_back(X);
    res += X;
    X = X * X % M;
  }
  N -= rec.size();
  if (N == 0) {
    printf("%lld\n", res);
    return 0;
  }
  ll freq = cyc.size();
  vector<ll> sum(freq + 1, 0);
  for (int i = 0; i < freq; ++i) {
    sum[i + 1] = sum[i] + cyc[i];
  }
  ll q = N / freq;
  ll r = N % freq;
  res += sum[freq] * q + sum[r];
  printf("%lld\n", res);
}